keras_nlp/samplers/beam_sampler.py

```
@@ -0,0 +1,218 @@
+from keras_nlp.samplers.sampler import call_args_docstring
+from keras_nlp.utils.python_utils import format_docstring
+
+
+@format_docstring(
+    base_sampler_args=base_sampler_args_docstring, call_args=call_args_docstring
+)
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class BeamSampler(Sampler):
+    """Beam Sampler class.
+
+    This sampler implements beam search algorithm. At each time-step, beam
+    search keeps the beams (sequences) of the top `num_beams` highest
+    accumulated probabilities, and uses each one of the beams to predict
+    candidate next tokens.
+
+    Args:
+        num_beams: int. The number of beams that should be kept at each
+            time-step. `num_beams` should be strictly positive.
+        {{base_sampler_args}}
+
+    Call Args:
+        {{call_args}}
+
+    Examples:
+    ```python
+    VOCAB_SIZE = 10
+
+    # Create a dummy model to predict the next token.
+    model = keras.Sequential(
+        [
+            keras.Input(shape=[None]),
+            keras.layers.Embedding(
+                input_dim=VOCAB_SIZE,
+                output_dim=16,
+            ),
+            keras.layers.Dense(VOCAB_SIZE, activation="softmax"),
+        ]
+    )
+
+    # Define a function that outputs the next token's probability for each token
+    # in the input sequence.
+    def token_probability_fn(inputs, mask):
+        return model(inputs)
+
+    prompt = tf.fill((8, 1), 1)
+
+    sampler = keras_nlp.samplers.BeamSampler(num_beams=3)
+    # Print the generated sequence (token ids).
+    print(sampler(prompt, token_probability_fn, max_length=10))
+    ```
+    """
+
+    def __init__(
+        self,
+        num_beams=5,
+        jit_compile=True,
+        run_eagerly=False,
+    ):
+        self.num_beams = num_beams
+        super().__init__(jit_compile=jit_compile, run_eagerly=run_eagerly)
+
+    def get_next_token(self, next_token_probs):
+        # Beam search overrides the whole `sample` method.
+        pass
```

This should raise an error I suppose?